test/new-e2e/utils/e2e/client/docker.go

```diff
@@ -0,0 +1,139 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package client
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/DataDog/test-infra-definitions/components/datadog/agent/docker"
+	"github.com/DataDog/test-infra-definitions/components/os"
+	"github.com/docker/cli/cli/connhelper"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/client"
+	"github.com/docker/docker/pkg/stdcopy"
+	"github.com/stretchr/testify/require"
+)
+
+var _ clientService[docker.ClientData] = (*Docker)(nil)
+
+// A docker client that is connected to an [docker.Deamon].
+//
+// [docker.Deamon]: https://pkg.go.dev/github.com/DataDog/test-infra-definitions@main/components/datadog/agent/docker#Deamon
+type Docker struct {
+	agent *AgentCommandRunner
+	*UpResultDeserializer[docker.ClientData]
+	t                  *testing.T
+	client             *client.Client
+	agentContainerName string
+	os                 os.OS
+}
+
+// Create a new instance of Docker
+func NewDocker(daemon *docker.Daemon) *Docker {
+	dockerInstance := &Docker{
+		agentContainerName: daemon.GetAgentContainerName(),
+		os:                 daemon.GetOS(),
+	}
+	dockerInstance.UpResultDeserializer = NewUpResultDeserializer[docker.ClientData](daemon, dockerInstance)
+	return dockerInstance
+}
+
+//lint:ignore U1000 Ignore unused function as this function is call using reflection
+func (docker *Docker) initService(t *testing.T, data *docker.ClientData) error {
```

Comment:
<!-- Praises highlight something positive. Try to leave at least one of these comments per review. Do look for something to sincerely praise. -->
**👏 praise**
This function is cool